File: WEB-INF/classes/com/x10clab/jaring/Jaring.java
package com.x10clab.jaring;

import java.io.File;
import java.io.FileInputStream;
import java.util.Properties;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.Statement;
import java.sql.SQLException;
import javax.servlet.http.Cookie;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.ServletContext;

import com.jolbox.bonecp.BoneCP;
import com.jolbox.bonecp.BoneCPConfig;

import com.alibaba.fastjson.JSONArray;
import com.alibaba.fastjson.JSONObject;

public class Jaring {
	public static String	_title			= "x10c Lab - Jaring Framework";
	public static String	_name			= "jaring";
	public static String	_path			= "/jaring";
	public static String	_path_mod		= "/module";
	public static int		_content_type	= 0;
	public static int		_paging_size	= 50;
	public static String	_db_class		= "";
	public static String	_db_url			= "";
	public static String	_db_user		= "";
	public static String	_db_pass		= "";
	public static int		_db_pool_min	= 10;
	public static int		_db_pool_max	= 100;
	private static BoneCP	_db_pool		= null;

	/*
		Cookies values.
		Variables that will be instatited when calling getCookiesValue.
	*/
	public static long		_c_uid			= 0;

	/*
		Initialize database connection pooling
	*/
	public static void initDBPool (HttpServletRequest request)
	{
		if (null != _db_pool) {
			return;
		}

		try {
			BoneCPConfig config	= new BoneCPConfig();

			Class.forName (Jaring._db_class);

			config.setJdbcUrl (Jaring._db_url);
			config.setUsername (Jaring._db_user);
			config.setPassword (Jaring._db_pass);
			config.setMinConnectionsPerPartition (Jaring._db_pool_min);
			config.setMaxConnectionsPerPartition (Jaring._db_pool_max);
			config.setPartitionCount (1);

			_db_pool = new BoneCP (config);

		} catch (Exception e) {
			e.printStackTrace ();
			return;
		}
	}

	/*
		Get one database connection from connection-pool.
	*/
	public static Connection getConnection (HttpServletRequest request)
	{
		try {
			if (null == _db_pool) {
				Jaring.initDBPool (request);
			}

			return Jaring._db_pool.getConnection ();
		} catch (Exception e) {
			e.printStackTrace ();
			return null;
		}
	}

	/*
		Get cookies values.
	*/
	public static void getCookiesValue (HttpServletRequest request)
	{
		Cookie[]	_cookies	= request.getCookies ();

		if (null != _cookies) {
			for (int i = 0; i < _cookies.length; i++) {
				String	c_name = _cookies[i].getName ();
				if (c_name.equalsIgnoreCase (Jaring._name +".user.id")) {
					Jaring._c_uid = Integer.parseInt (_cookies[i].getValue ());
				}
			}
		}
	}

	/*
		Get module permission.
	*/
	public static int getModulePermission (HttpServletRequest request, int menu_id)
	{
		Connection			con = null;
		PreparedStatement	ps	= null;
		ResultSet			rs	= null;
		String				q	= "";
		int					i	= 0;
		int					perm= 0;

		try {
			con = Jaring.getConnection (request);

			/* get cookie value for user ID */
			Jaring.getCookiesValue (request);

			q	="	select	A.permission"
				+"	from	_group_menu	A"
				+"	,		_user_group	B"
				+"	where	A._menu_id	= ?"
				+"	and		A._group_id	= B._group_id"
				+"	and		B._user_id	= ?"
				+"	order by A.permission desc";

			ps	= con.prepareStatement (q);
			i	= 1;
			ps.setLong (i++	, menu_id);
			ps.setLong (i++	, Jaring._c_uid);

			rs	= ps.executeQuery ();

			if (rs.next ()) {
				perm = rs.getInt ("permission");
			}

			rs.close ();
			ps.close ();
			con.close ();
		} catch (Exception e) {
			/**/
		} finally {
			return perm;
		}
	}

	/*
		Get parameter in number format.
	*/
	public static int getIntParameter (HttpServletRequest request
										, String paramName
										, int defaultValue) {
		String	paramString = request.getParameter (paramName);
		int		paramValue	= defaultValue;
		try {
			paramValue = Integer.parseInt (paramString);
		} catch (NumberFormatException nfe) { // Handles null and bad format
			paramValue = defaultValue;
		}
		return (paramValue);
	}

	/*
		Initialize application
	*/
	public static void init (ServletContext application, HttpServletRequest request)
	{
		try {
			/* Load application configuration */
			String			app_conf		= application.getRealPath ("WEB-INF"+ File.separator +"app.conf");
			FileInputStream	fis				= new FileInputStream (app_conf);
			Properties		props			= new Properties ();

			props.load (fis);

			Jaring._path		= application.getContextPath ();
			Jaring._title		= props.getProperty ("app.title");
			Jaring._name		= props.getProperty ("app.name");
			Jaring._path_mod	= props.getProperty ("app.module.dir");
			Jaring._content_type= Integer.parseInt (props.getProperty ("app.content.type"));
			Jaring._paging_size	= Integer.parseInt (props.getProperty ("app.paging.size"));
			Jaring._db_class	= props.getProperty ("db.class");
			Jaring._db_url		= props.getProperty ("db.url");
			Jaring._db_user		= props.getProperty ("db.username");
			Jaring._db_pass		= props.getProperty ("db.password");
			Jaring._db_pool_min	= Integer.parseInt (props.getProperty ("db.pool.min"));
			Jaring._db_pool_max	= Integer.parseInt (props.getProperty ("db.pool.max"));

			/* Always replace global config variables, in case user change it on the fly */
			application.setAttribute ("app.title"		, Jaring._title);
			application.setAttribute ("app.name"		, Jaring._name);
			application.setAttribute ("app.path"		, Jaring._path);
			application.setAttribute ("app.module.dir"	, Jaring._path_mod);
			application.setAttribute ("app.content.type", Jaring._content_type);
			application.setAttribute ("app.paging.size"	, Jaring._paging_size);
			application.setAttribute ("db.class"		, Jaring._db_class);
			application.setAttribute ("db.url"			, Jaring._db_url);
			application.setAttribute ("db.username"		, Jaring._db_user);
			application.setAttribute ("db.password"		, Jaring._db_pass);
			application.setAttribute ("db.pool.min"		, Jaring._db_pool_min);
			application.setAttribute ("db.pool.max"		, Jaring._db_pool_max);

			/* Initialize database connection-pool */
			initDBPool (request);
		} catch (Exception e) {
			e.printStackTrace ();
		}
	}
}

File: WEB-INF/classes/com/x10clab/jaring/Jaring.php
<?php
class SafePDO extends PDO
{
	public static function exception_handler($exception)
	{
		// Output the exception details
		die('Uncaught exception: '. $exception->getMessage());
	}

	public function __construct($dsn, $username='', $password='', $driver_options=array())
	{
		// Temporarily change the PHP exception handler while we . . .
		set_exception_handler(array(__CLASS__, 'exception_handler'));

		// . . . create a PDO object
		parent::__construct($dsn, $username, $password, $driver_options);

		// Change the exception handler back to whatever it was before
		restore_exception_handler();
	}
}

class Jaring
{
	public static $MSG_SUCCESS_UPDATE	= 'Data has been updated.';
	public static $MSG_SUCCESS_CREATE	= 'New data has been created.';
	public static $MSG_SUCCESS_DESTROY	= 'Data has been deleted.';
	public static $MOD_INIT				= '/init';

	public static $_ext				= '.php';
	public static $_title			= 'x10c Lab - Jaring Framework';
	public static $_name			= 'jaring';
	public static $_path			= '/stp';
	public static $_path_mod		= '/module';
	public static $_mod_init		= '';
	public static $_content_type	= 0;
	public static $_menu_mode		= 1;
	public static $_paging_size		= 50;
	public static $_db_class		= '';
	public static $_db_url			= '';
	public static $_db_user			= '';
	public static $_db_pass			= '';
	public static $_db_pool_min		= 0;
	public static $_db_pool_max		= 100;
	public static $_db				= null;

	/*
		Cookies values.
		Variables that will be instantiated when calling getCookiesValue.
	*/
	public static $_c_uid			= 0;
	public static $_c_username		= 'Anonymous';

	public static function initDB ()
	{
		self::$_db = new SafePDO (self::$_db_url, self::$_db_user, self::$_db_pass);
		self::$_db->setAttribute(PDO::ATTR_ERRMODE, PDO::ERRMODE_EXCEPTION);
	}
	
	public static function getConnection ()
	{
		if (self::_db == null) {
			self::initDB ();
		}
	}
	
	public static function getCookiesValue ()
	{
		$ckey = 'user_id';
		if (isset ($_COOKIE[$ckey])) {
			self::$_c_uid = (int) $_COOKIE[$ckey];
		}

		$ckey = 'user_name';
		if (isset ($_COOKIE[$ckey])) {
			self::$_c_username = $_COOKIE[$ckey];
		}
	}
	
	public static function init ()
	{
		$f_app_conf	= APP_PATH ."/WEB-INF/app.conf";

		$app_conf = parse_ini_file ($f_app_conf);

		self::$_title			= $app_conf['app.title'];
		self::$_name			= $app_conf['app.name'];
		self::$_ext				= $app_conf['app.extension'];
		self::$_path			= $app_conf['app.path'];
		self::$_path_mod		= $app_conf['app.module.dir'];
		self::$_mod_init		= self::$_path . self::$_path_mod . self::$MOD_INIT . self::$_ext;
		self::$_content_type	= $app_conf['app.content.type'];
		self::$_menu_mode		= $app_conf['app.menu.mode'];
		self::$_paging_size		= $app_conf['app.paging.size'];
		self::$_db_class		= $app_conf['db.class'];
		self::$_db_url			= $app_conf['db.url'];
		self::$_db_user			= $app_conf['db.username'];
		self::$_db_pass			= $app_conf['db.password'];
		self::$_db_pool_min		= $app_conf['db.pool.min'];
		self::$_db_pool_max		= $app_conf['db.pool.max'];

		Jaring::getCookiesValue ();
	}
}
?>